zebra-chain/src/types/amount.rs

```diff
@@ -9,6 +9,102 @@ use std::{
 #[derive(Debug, Eq, PartialEq, Clone, Copy, Serialize, Deserialize)]
 pub struct Amount<C = NegativeAllowed>(i64, PhantomData<C>);
 
+impl<C1, C2> std::ops::Add<Amount<C2>> for Amount<C1>
+where
+    C1: AmountConstraint,
+    C2: AmountConstraint,
+{
+    type Output = Option<Amount<C1>>;
+
+    fn add(self, rhs: Amount<C2>) -> Self::Output {
+        let value = self.0 + rhs.0;
+        value.try_into().ok()
+    }
+}
+
+impl<C1, C2> std::ops::Add<Amount<C2>> for Option<Amount<C1>>
+where
+    C1: AmountConstraint,
+    C2: AmountConstraint,
+{
+    type Output = Option<Amount<C1>>;
+
+    fn add(self, rhs: Amount<C2>) -> Self::Output {
+        self.and_then(|this| this + rhs)
```

Comment:
I think I'm missing something -- where does this do the validation?  Shouldn't there be a call to `AmountConstraint::validate` somewhere in here?

Reply:
Its being handled by the `this + rhs` expr which is forwarding to the Add impl without `Option`s on either side of the expr

Reply:
Ah, that's quite nice